packages/ai-jsx/src/lib/tts/tts.tsx

```diff
@@ -0,0 +1,360 @@
+'use client';
+
+/**
+ * Defines a function that can be used to build a URL for a text-to-speech
+ * service. This URL will be used to retrieve an audio file that can be
+ * played by an HTML5 audio element.
+ */
+export type BuildUrlFunction = (provider: string, voice: string, rate: number, text: string) => string;
+
+/**
+ * Defines a function that can be used to retrieve an ephemeral access token
+ * for us with a text-to-speech service. Typically, this will be a
+ * fetch request to a server that will return an opaque token.
+ */
+export type GetTokenFunction = (provider: string) => Promise<string>;
+
+/**
+ * Defines a base class for text-to-speech services. This class provides
+ * a common interface for text-to-speech services, as well as some basic
+ * infrastructure for playing audio using the HTML5 audio element.
+ */
+export class TextToSpeechBase {
+  protected audio: HTMLAudioElement;
+  /**
+   * The time when the play() method was first called.
+   */
+  protected playMillis: number = 0;
+  /**
+   * The latency between when the play() method is called and when the audio starts playing.
+   */
+  public latency: number = 0.0;
+  /**
+   * Called when the generated audio has started playing.
+   */
+  public onPlaying?: () => void;
+  /**
+   * Called when the generated audio has finished playing out.
+   */
+  public onComplete?: () => void;
+
+  constructor(protected readonly name: string) {
+    this.audio = new Audio();
+    this.audio.onplay = () => console.log(`[${this.name}] tts playing`);
+    this.audio.onpause = () => console.log(`[${this.name}] tts paused`);
+    this.audio.onloadstart = () => console.log(`[${this.name}] tts loadstart`);
+    this.audio.onloadeddata = () => console.log(`[${this.name}] tts loadeddata`);
+    this.audio.oncanplay = () => console.log(`[${this.name}] tts canplay`);
+    this.audio.onplaying = () => {
+      console.log(`[${this.name}] tts playing`);
+      if (this.playMillis) {
+        this.latency = Math.floor(performance.now() - this.playMillis);
+        console.log(`[${this.name}] tts play latency: ${this.latency} ms`);
+      }
+      this.onPlaying?.();
+    };
+  }
+  /**
+   * Converts the given text to speech and plays it using the HTML5 audio element.
+   * This method may be called multiple times, and the audio will be played serially.
+   * Generation may be buffered, use the flush() method to indicate all text has been provided.
+   */
+  play(_text: string) {
+    throw new Error('Method not implemented.');
```

Comment:
Any reason not to use `abstract` here?

https://www.perplexity.ai/search/e3aee5ee-93c6-42fc-b980-9d8d8c63c0d3?s=u